src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -739,6 +740,93 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomStringFromDataset : public Generator<std::string> {
+
+public:
+    RandomStringFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()} {
+        loadDataset();
+    }
+
+private:
+    void loadDataset() {
+        {
+            if (_path.empty()) {
+                BOOST_THROW_EXCEPTION(
+                    InvalidValueGeneratorSyntax("ChooseFromDataset requieres non-empty path"));
+            }
+
+            /* If path the doesn't exist in memory we load it.
+               To avoid reading and storing the datasets multiple times
+               we check if the path already exists in the map (as key). If
+               it doesn't exist, we open the file and store its contents.
+               If the number of items is more than 0, it means the data is
+               loaded, so we can skip everything, including the mutex.
+            */
+
+            if (_all_datasets.count(_path) > 0) {
+                return;
+            }
+
+            {
+                std::mutex _dataset_mutex;
+                std::lock_guard<std::mutex> lk(_dataset_mutex);
+
+                if (_all_datasets.count(_path) == 0) {
+                    std::string currentLine;
+                    std::ifstream ifs;
+
+                    /* Add a first empty vector as soon as possible
```

Comment:
This only works if we have multiple locks, right? There's the `_all_datasets` lock that should be locked when adding a new dataset, and a dataset-specific lock that should be used when populating a specific dataset. If we hold the `_all_datasets` lock while populating a specific dataset, then all threads are blocked on every dataset load, even if they don't care about that dataset.

It seems like we should have the `_all_datasets` lock that is used when adding a new dataset vector, and a dataset-specific lock used to populate that dataset. I imagine we'd take the `_all_datasets` lock, add the vector and dataset-specific lock as a pair, release the lock, take the dataset-specific lock, and populate the vector, then release it.

Also, we might want to consider a reader-writer lock system like the orchestrator uses [here](https://github.com/mongodb/genny/blob/24e68966b9846d0ba0615c8bc4503008e04b58c1/src/gennylib/src/Orchestrator.cpp#L37).

Reply:
Hello @aggrand. That makes sense, but how can I create locks at runtime? I don't know in advance how many and which datasets are going to be loaded, so I cannot define in the code the lock names. Maybe I am misunderstanding something?

Reply:
I think we should only ever have one thread populating the vector. My thought in pseudo-code is we should have something like:
```
results = []

lock(_map)
if(path in _map)
  // someone else is already loading, nothing to do
  return
else
 // this thread is going to load the dataset
 _map.add(path, results)
unlock(_map)

for(item in file) {
  results.add(item)
}
```

If another two threads are loading the same dataset, the first one to enter here would hit the map.add path, and the second would immediately bail out. If two threads are loading different datasets, both get a unique results list that they are populating after unlocking the map. Let me know if I'm missing something here; I think if this is the case we only need a single mutex.

Reply:
Correct me if I'm wrong, but with the above pseudocode, a thread might take the lock, populate the map with the empty vector, drop the lock, then start populating the vector. Another thread might come along, take the lock, see that the map contains the vector, then assume it's populated with no way of knowing that the population is in-progress, then try to use the vector while the population is incomplete.

Reply:
Hence the two levels of locking, for the multi-dataset map and for the dataset.

Reply:
@aggrand I don't think we ever have reading and writing at the same time of the actual datasets. We should write them during initialization, and read them during workload execution. 

@brad-devlugt I think your code is technically correct, but a bit scary :) If someone bungled the early return such that it just exited that scope rather than the function, we would have multiple threads installing the dataset at the same time. I think it makes sense to have a lock guard around updating the items in the dataset. Does that make sense? I might go a step further and say the dataset vector should be built locally, and then moved into the _all_datasets maps. That way, if we have a lock on it, but we still somehow try to add it twice (bug), we end up with the right value.

Reply:
i.e. we have two resources: a global registry of datasets and the contents of a specific dataset. We can ensure safety at the cost of efficiency by guarding them both with the same lock, but if we want both resources to be interact-able independently it seems like we need a lock for each.

Reply:
@aggrand yes.

Reply:
Yeah @aggrand that's a concern if the evaluate methods can be called in other threads at the same time as the constructor. Does genny construct the generators lazily or up front? Is the constructing stage multi or single threaded?

Reply:
Then, can it be something like this? This way the threads will be waiting on LockOne (checking is the path's vector is there) for a short time, just until the first thread adds the empty vector. This way, we won't have more than one thread loading the data, and the time waiting on the lockone will be short.

```
   {      
   GetLockOne()
   if (_all_datasets.count(_path) > 0) {
                return;
   }

            {
                GetLockTwo()
                if (_all_datasets.count(_path) == 0) {
                    std::string currentLine;
                    std::ifstream ifs;

                    /* Add a first empty vector as soon as possible
                       so new threads that are being created do not
                       have to wait to acquire the mutex */

                    _all_datasets[_path] = std::vector<std::string>();
		    UnlockLockOne()
...
```

Reply:
@brad-devlugt I believe the generators are constructed upfront because they are part of the actor classes, e.g. [here](https://github.com/mongodb/genny/blob/24e68966b9846d0ba0615c8bc4503008e04b58c1/src/cast_core/src/Loader.cpp#L59), and as of [this](https://github.com/mongodb/genny/pull/562
) commit constructing actors is indeed [parallelized](https://github.com/mongodb/genny/blob/24e68966b9846d0ba0615c8bc4503008e04b58c1/src/gennylib/src/context.cpp#L83).

@miguelangelnieto I think your code there has the lock two attained inside the scope of lock one, based on the indent / still open paranetheses. I think we should really treat adding to the registry and populating the vector as basically two independent operations, which are sequential

Reply:
thanks - so it sounds like, if we construct them up-front, we only need to lock the map. This is since no thread would try to use the vector in evaluate() until all the generators are constructed. And if we are constructing actors in parallel, that has the nice benefit of letting us load different datasets in parallel.

Reply:
We construct them up front, but as I see it if lock only the map during the whole vector population and don't get that parallelism benefit. And if we want to prevent multiple writers to the same vector we need to lock it, right?

Or wait actually we could set a flag when we are the thread to create the vector to also say we are the thread to populate it, to associate the two events. So any thread that doesn't create the mapping just returns?

Reply:
Thank you. So, based on your conversation with Brad, and based in his pseudocode, I understand that this is what we want. Lock the map, and then only the first thread will populate the vector. Let me know if my understanding is not correct.

```
std::vector<std::string> data;

{
    std::lock_guard<std::mutex> lk(_dataset_mutex);    
    if (_all_datasets.count(_path) > 0) {
        return;
    }

    else {
    _all_datasets[_path] = data;
    }
}

std::string currentLine;
std::ifstream ifs;

ifs.open(_path, std::ifstream::in);
if (ifs.is_open()) {
    while (std::getline(ifs, currentLine)) {
        if (!currentLine.empty()) {
            _all_datasets[_path].push_back(currentLine);
        }
    }
}
```

Reply:
yup I think the above sounds good! Maybe a comment on the return to make it clearer that the return MUST happen

Reply:
That quote hit a formatting problem. The exception is std::vector\<bool\>. I'll see if I can fix it.

Reply:
Yeah I agree, the elements can be modified concurrently, this is why we can push to the vector via the pointer. I think the lookup via `operator []` isn't thread-safe for associative containers, however. Also from https://en.cppreference.com/w/cpp/container#Thread_safety:
>  and, except in associative containers, operator[], behave as const for the purposes of thread safety

I think with unordered_map, for example, an insert could trigger a re-hashing and then one thread could be in the middle of that when the read comes in.

Reply:
Ah, okay. And better safe than sorry here. Thanks Brad.

Reply:
No problem! I had to debug a really tricky problem once before that ended up being related to the thread-safety of `operator []`.


@miguelangelnieto I don't think it is worth 1-2 weeks of time, so my vote is to either time-box it at an hour or move forward with your proposed fix and keep the pattern in mind as an option for next time. The time-savings with the pattern comes from catching errors at compile time vs runtime/code review, and in time savings when working in the area of code next time around.